addons/sale_purchase_project/models/account_move_line.py

```diff
@@ -0,0 +1,17 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    reinvoiced_so_id = fields.Many2one('sale.order')
+
+    def _sale_determine_order(self):
+        mapping_from_invoice = super()._sale_determine_order()
+        for move_line in self:
+            sale_order = mapping_from_invoice.get(move_line.id)
+            if sale_order:
+                move_line.reinvoiced_so_id = sale_order
```

Comment:
Don't you need to clear `move_line.reinvoiced_so_id` if `sale_order` is falsy then?
```suggestion
            move_line.reinvoiced_so_id = mapping_from_invoice.get(move_line.id, self.env['sale.order'])
```

Reply:
I don't think this is necessary because we want to keep track of the last SO/SOL the invoice line was re-invoiced on. If no such exists, we still keep the last one in memory. If the user really wants to get rid of, he can easily remove the re-invoiced SO/SOL and `reinvoiced_so_id`will be empty then :)